lib/src/package_analysis/common.dart

```diff
@@ -1,12 +1,204 @@
+/// pinning the dependency [name] to version [version].
+Future<void> fetchPackageWithPointer({
+  required String name,
+  required String version,
+  required String destination,
+  required bool wipeTarget,
+}) async {
+  // delete the target directory, if it exists and wipe is enabled
+  if (wipeTarget && await Directory(destination).exists()) {
+    await Directory(destination).delete(recursive: true);
+  }
+
+  // construct pubspec for the pointer package
+  final pointerPubspec = <String, dynamic>{
+    'name': 'dummy_package',
+    'environment': {
+      'sdk': '>=2.12.0 <3.0.0',
+    },
+    'dependencies': {},
+  };
+  pointerPubspec['dependencies'][name] = version;
+
+  // if pubspec.yaml exists, delete it and create an empty file
+  final pubspecFile = File(path.join(destination, 'pubspec.yaml'));
+  if (await pubspecFile.exists()) {
+    await pubspecFile.delete();
+  }
+  await pubspecFile.create(recursive: true);
+
+  // write pubspec to disk
+  await pubspecFile.writeAsString(json.encode(pointerPubspec));
+
+  // fetch dependencies (the only non-transitive dependency is [name])
+  await fetchDependencies(destination);
+}
+
+/// Fetches dependencies in [destination].
+Future<void> fetchDependencies(String destination) async {
+  final result = await Process.run(
+    Platform.resolvedExecutable,
+    ['pub', 'get'],
+    workingDirectory: destination,
+  );
+  if (result.exitCode != 0) {
+    throw ProcessException(
+      Platform.resolvedExecutable,
+      ['pub', 'get'],
+      result.stderr.toString(),
+      result.exitCode,
+    );
+  }
+}
+
+/// Given the location of a package and the name of one of its dependencies,
+/// return the path of the dependency, or null if it cannot be resolved.
+/// Ensure that the dependencies of the target package are fetched.
+Future<String?> getDependencyDirectory(
+  PackageAnalysisContext packageAnalysisContext,
+  String packageLocation,
+  String dependencyName,
+) async {
+  final dependencyUri = Uri.parse('package:$dependencyName/');
+  final dependencyFilePath = packageAnalysisContext.uriToPath(dependencyUri);
+  print('path for $dependencyName in $packageLocation is $dependencyFilePath');
+  return dependencyFilePath == null ? null : path.dirname(dependencyFilePath);
+}
+
+/// Recursively fetch all the files in the given folder
+List<resource.File> getAllFiles(resource.Folder folder) {
+  final files = <resource.File>[];
+  for (final child in folder.getChildren()) {
+    if (child is resource.File) {
+      files.add(child);
+    } else if (child is resource.Folder) {
+      files.addAll(getAllFiles(child));
+    } else {
+      throw StateError(
+          'Failed to determine subclass of Resource ${child.path}.');
+    }
+  }
+  return files;
+}
+
+/// Fetch all the hosted dependencies used by a package at a given location
+Future<Map<String, HostedDependency>> getHostedDependencies(
+    String targetPackageDir) async {
```

Comment:
Just a suggestion, feel free to ignore.
```suggestion
Future<Map<String, HostedDependency>> getHostedDependencies(
  String targetPackageDir,
) async {
```